### PR TITLE
Bug 1568479 - close HTTP response bodies when caller can't/shouldn't

### PR DESCRIPTION
**Pull request by petemoore**
See [bug 1568479](https://bugzil.la/1568479) for context. Thanks!

### CODE REVIEW

**Comment by petemoore** [comment]
Hi guys, on reflection this will require a little rethinking. The issue is that normally, we'd expect the caller to only close the response body if an error didn't occur, so for non-retrying code like this:

```go
resp, err := http.Get("https://twitter.com/watchallirish/status/1153601865713639424")
if err != nil {
  // return an error / panic / call log.Fatal / fail test / etc
}
defer resp.Body.Close()
```

we currently recommend that it should be rewritten like this:

```go
resp, _, err := httpbackoff.Get("https://twitter.com/watchallirish/status/1153601865713639424")
if err != nil {
  // return an error / panic / call log.Fatal / fail test / etc
}
defer resp.Body.Close()
```

The problem with this, however, is that when using the `httpbackoff` library, the caller may want access to read the http response body content, if there is any, __also__ when an error occurs (e.g. an http response with status code 400 and response body containing information about the reason for failure).

Therefore, it is probably better that we instead advise the caller that the response body should be closed if the http response is not `nil`, i.e. like this:

```
resp, attempts, err := httpbackoff.Get("https://twitter.com/watchallirish/status/1153601865713639424")

if resp != nil {
  defer resp.Body.Close()
}

if err != nil {
  // return an error / panic / call log.Fatal / fail test / etc
}
```

This way, the decision about whether to close the response body is not made based on whether an error is returned, since the error could be that the http request could not be made (so there is no response body) or it could be that a http request was made, and there was a response, it just had a status code that indicates there was an error.

An alternative is that we leave it like it is, and advise users to check if the returned error is of type `httpbackoff.BadHttpResponseCode`, and if so, that the response body can be retrieved from there.

Note, ideally we wouldn't be caching response bodies in memory at all, and would use e.g. temporary files if we really need to, or would use memory up to a limit of e.g. 64KB, and if the response is larger, we use using a temporary file instead.

File: httpbackoff.go [hunk 1]
```diff
@@ -67,16 +67,24 @@ func (err BadHttpResponseCode) Error() string {
...
+// httpCall should be a function that performs the http operation. Errors that
+// should cause retries should be returned as tempError. Permanent errors that
+// should not result in retries should be returned as permError. Retries are
+// performed using the exponential backoff algorithm from the
+// github.com/cenkalti/backoff package. Retry treats HTTP 5xx status codes as a
+// temporary error, and any other non-2xx HTTP status codes as a permanent
+// error. Thus httpCall does not need to handle the HTTP status code of resp,
 // since Retry will take care of it.
 //
+// if httpCall returns either a temporary or permanent error, Retry will *not*
```
Comment (milescrabill) [inline httpbackoff.go]: It's worth considering whether we should create a `type HttpCall func() (resp *http.Response, tempError error, permError error)` and document its expected behavior as a comment on the type.

I think we can clean up / simplify our phrasing here. If this the following properties are true, we can recycle the phrasing used in the `http` package:

> "When err is nil, resp always contains a non-nil resp.Body. Caller should close `resp.Body` when done reading from it."

How about this phrasing:

 > When tempError and permError are nil, Retry expects resp to contain a non-nil resp.Body.
> When tempError or permError are non-nil, Retry expects httpCall to close resp.Body.
> When resp contains a non-2XX response code, Retry returns an error and closes resp.Body.
> When resp contains a 2XX response code, Retry expects its caller to close resp.Body when done reading from it.

Comment (milescrabill) [reply httpbackoff.go]: The amount of digesting I had to do to understand this behavior worries me a bit. (and even then I might have something wrong!)

File: httpbackoff.go [hunk 2]
```diff
@@ -102,19 +110,28 @@ func (httpRetryClient *Client) Retry(httpCall func() (resp *http.Response, tempE
 		}
 		// now check if http response code is such that we should retry [500, 600)...
 		if respCode := response.StatusCode; respCode/100 == 5 {
+			// If this is not the last attempt, the request will be retried and
+			// the response body should be closed. If this is the last attempt,
+			// Retry will return an error, so caller isn't responsible for
+			// closing response body. Therefore, in either case, response body
+			// needs to be closed.
+			defer response.Body.Close()
```
Comment (djmitche) [inline httpbackoff.go]: I'm not seeing the connection between this comment and this line of code, and neither seems aligned with the header comment.  The code doesn't seem concerned with the distinction of last vs. non-final retry.  And, if this was a 5xx error, then we expected the caller to close the body, right?

Also, `defer` seems to result in ignoring an error here.  Is that intentional?

**Comment by petemoore** [comment]
Closing for now as I'm not actively working on it.

**Pull request closed without merging**